momento-functions/examples/turbopuffer-recommended-articles.rs

```diff
@@ -0,0 +1,453 @@
+//!   --cache-name "$MOMENTO_CACHE_NAME" \
+//!   --name turbopuffer-recommended-articles \
+//!   --wasm-file /path/to/this/compiled/turbopuffer_search_articles.wasm \
+//!   -E OPENAI_KEY="$OPENAI_KEY" \
+//!   -E TURBOPUFFER_ENDPOINT="$TURBOPUFFER_ENDPOINT" \
+//!   -E TURBOPUFFER_API_KEY="$TURBOPUFFER_API_KEY"
+//!
+//! # Find some recommended articles!
+//! curl --silent \
+//! https://api.cache.$MOMENTO_CELL_HOSTNAME/functions/$MOMENTO_CACHE_NAME/turbopuffer-recommended-articles \
+//!  -H "authorization: $MOMENTO_API_KEY" \
+//!  -H "Content-Type: application/json" \
+//!  -d '{"article_ids": ["4484641135897252353", "6703952717813182352", "4965577503700120031"]'  | jq
+//! ```
+
+use std::time::Duration;
+
+use log::LevelFilter;
+
+use momento_functions::{WebError, WebResponse, WebResult};
+use momento_functions_host::{cache, encoding::Json, web_extensions::headers};
+use momento_functions_log::LogMode;
+use serde::{Deserialize, Serialize};
+use serde_json::{Value, json};
+
+#[derive(Deserialize, Debug)]
+struct Request {
+    article_ids: Vec<String>,
+    topk: Option<usize>,
+}
+
+#[derive(Deserialize, Debug)]
+struct QueryResponse {
+    rows: Vec<QueryRow>,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+struct QueryRow {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(alias = "$dist")]
+    dist: Option<f32>,
+    id: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    vector: Option<Vec<f32>>,
+    #[serde(rename = "metadata$title")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    metadata_title: Option<String>,
+    #[serde(rename = "metadata$link")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    metadata_link: Option<String>,
+}
+
+// Default to 5 minute caching time for queries in Momento
+const DEFAULT_TTL_SECONDS: u64 = 300;
+// Filter out articles that aren't similar enough
+const MAXIMUM_COSINE_DISTANCE: f32 = 0.6;
+
+momento_functions::post!(get_recommended_articles);
+fn get_recommended_articles(Json(request): Json<Request>) -> WebResult<WebResponse> {
+    let headers = headers();
+    setup_logging(&headers)?;
+
+    // These are passed in as environment variables when creating the function
+    let turbopuffer_api_key = format!(
+        "Bearer {}",
+        std::env::var("TURBOPUFFER_API_KEY").unwrap_or_default()
+    );
+    let turbopuffer_endpoint = std::env::var("TURBOPUFFER_ENDPOINT").unwrap_or_default();
+    let ttl_seconds = std::env::var("TTL_SECONDS")
+        .unwrap_or(DEFAULT_TTL_SECONDS.to_string())
+        .parse::<u64>()
+        .unwrap_or(DEFAULT_TTL_SECONDS);
+    // Default to top 5 results if not provided
+    let ttl = Duration::from_secs(ttl_seconds);
+
+    let Request { article_ids, topk } = request;
+
+    // Default to top 10 articles if not provided
+    let topk = topk.unwrap_or(10);
+
+    // Get the embeddings from our provided articles
+    let embeddings = get_cached_article_embeddings(
```

Comment:
`get_article_embeddings` - since we first check the cache and fall back to turbopuffer. The name is slightly misleading like this.